src/libxtp/gaussianwriter.cc

```diff
@@ -67,9 +67,22 @@ std::string GaussianWriter::reorderedMOCoefficients(
 }
 
 std::string GaussianWriter::densityMatrixToString(const Orbitals& orbitals,
-                                                  const QMState& state) const {
+                                                  const QMState& state,
+                                                  bool diff2gs) const {
   OrbReorder reorder(gaussianOrder, gaussianMultipliers, true);
-  Eigen::MatrixXd density = orbitals.DensityMatrixFull(state);
+
+  Eigen::MatrixXd density;
+  if (state.Type().isExciton() && diff2gs) {
+    std::array<Eigen::MatrixXd, 2> DMAT =
+        orbitals.DensityMatrixExcitedState(state);
+    density = DMAT[1] - DMAT[0];
+  } else if ((state.Type().isKSState() || state.Type().isPQPState()) &&
+             diff2gs) {
+    density = orbitals.DensityMatrixKSstate(state);
```

Comment:
please create a function `DifferenceDensityMatrix` or so and then `if` only on `diff2gs`

Reply:
and a test would be helpful

Reply:
Coming soon

Reply:
Or not, it is covered in the test_cubefile_writer, let me know if you think that that isn't enough, but for me that is good enough